MovieTime/MovieTime/Service/MTParameterEncoding.swift

```diff
@@ -0,0 +1,375 @@
+    public static var `default`: URLEncoding { return URLEncoding() }
+
+    public static var customQueryString: URLEncoding { return URLEncoding(customize: .reflection) }
+
+    public let customize: Customize
+
+    public init(customize: Customize = .none) {
+        self.customize = customize
+    }
+
+    // MARK: Encoding
+
+    public func urlEncode(_ url: String, method: MTHTTPMethod, with parameters: MTParameters?) -> URL? {
+        var url = url
+        if customize == .reflection {
+            // Encodes parameters in URL
+            if method.encodesParametersInURL() {
+                url = url.reflectionEncode(parameters: parameters)
+            }
+        }
+
+        let encodedURL = url.addingPercentEncoding(withAllowedCharacters: NSCharacterSet.urlQueryAllowed) ?? url
+        return URL(string: encodedURL)
+    }
+
+    public func encode(_ urlRequest: URLRequest, with parameters: MTParameters?) -> URLRequest {
+        var urlRequest = urlRequest
+        guard let parameters = parameters else { return urlRequest }
+
+        if let method = MTHTTPMethod(rawValue: urlRequest.httpMethod ?? "GET"), method.encodesParametersInURL() {
+            guard let url = urlRequest.url else { return urlRequest }
+
+            if customize != .reflection,
+                var urlComponents = URLComponents(url: url, resolvingAgainstBaseURL: false),
+                !parameters.isEmpty {
+
+                let percentEncodedQuery = (urlComponents.percentEncodedQuery.map { $0 + "&" } ?? "") + query(parameters)
+                urlComponents.percentEncodedQuery = percentEncodedQuery
+                urlRequest.url = urlComponents.url
+            }
+        } else {
+            if urlRequest.value(forHTTPHeaderField: "Content-Type") == nil {
+                urlRequest.setValue("application/x-www-form-urlencoded; charset=utf-8", forHTTPHeaderField: "Content-Type")
+            }
+
+            urlRequest.httpBody = query(parameters).data(using: .utf8, allowLossyConversion: false)
+        }
+
+        return urlRequest
+    }
+
+    /// Creates percent-escaped, URL encoded query string components from the given key-value pair using recursion.
+    ///
+    /// - parameter key:   The key of the query component.
+    /// - parameter value: The value of the query component.
+    ///
+    /// - returns: The percent-escaped, URL encoded query string components.
+    public func queryComponents(fromKey key: String, value: Any) -> [(String, String)] {
+        var components: [(String, String)] = []
+
+        if let dictionary = value as? [String: Any] {
+            for (nestedKey, value) in dictionary {
+                components += queryComponents(fromKey: "\(key)[\(nestedKey)]", value: value)
+            }
+        } else if let array = value as? [Any] {
+            for value in array {
+                components += queryComponents(fromKey: "\(key)[]", value: value)
+            }
+        } else if let value = value as? NSNumber {
```

Comment:
when do you need to use NSNumber in this app?

Reply:
One reason is for type conversion. For example:
```
var someDouble = 3.14
var nsNumberDouble: NSNumber = NSNumber(value: someDouble)
var someFloat = nsNumberDouble.floatValue
```